### PR TITLE
Correct VS coil operation with HX

### LINKED ISSUE #6896: Example file 5Zone_Unitary_HXAssistedCoil shows very cold cooling coil outlet temperature in winter when the coil is off.
Issue overview
--------------
Review of results show that it appears that the cooling coil is on when not needed even though the coil report of sensible cooling rate shows 0 in winter. Report variables shown are for the Coil:Cooling:DX:VariableSpeed coil model. This cooling coil is wrapped with a HX so not sure if that has something to do with these results. All coils in this model are controlled to 17C in winter and 12C in summer so zone temperature control is poor at best. Error file also shows problems with max iterations and other obvious failure errors.

image

Update 2/13/2020:

Heating/Reheat coils appear undersized, however, the cooling coil outlet temps are low during the heating season when cooling coil should be off (so heating coil are probably sized correctly).

image

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version)
 - Version of EnergyPlus (if using an intermediate build, include SHA)
 - Unmethours link or helpdesk ticket number

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added (use example file 5Zone_Unitary_HXAssistedCoil.idf)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by rraustad**
Pull request overview
---------------------
 - Fixes #6896
 - Corrects very low HX outlet temperatures when variable speed coil is used with CoilSystem:Cooling:DX:HeatExchangerAssisted.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by rraustad** [comment]
This was a mistake in VS coil simulate argument list.

image

image

image

**Comment by nrel-bot-2** [comment]
@rraustad @lgentile it has been 28 days since this pull request was last updated.

**Comment by mitchute** [comment]
@rraustad was this fixed in #7886?

**Comment by rraustad** [comment]
HXssistedCoolingCoil is not one of the modified files in #7886.

**Comment by mitchute** [comment]
OK, thanks. It looks like the fix mentioned in the following comment must have been temporary.

https://github.com/NREL/EnergyPlus/pull/7886#issuecomment-606977460

**Comment by rraustad** [comment]
I fixed conflict and running unit tests now.

**Comment by rraustad** [comment]
I remember now, there were diff's in the HVACHXAssistedCoolingCoil file and to be sure the smaller heating coil could still meet the zone load I tested this fix over there but didn't keep the changes in that branch.

**Comment by rraustad** [comment]
Looks like I need to add a unit test.

File: src/EnergyPlus/UnitarySystem.cc [hunk 1]
```diff
@@ -7789,7 +7789,7 @@ namespace UnitarySystems {
                                                 Real64 &OnOffAirFlowRatio,     // ratio of heating PLR to cooling PLR (is this correct?)
                                                 Real64 const ZoneLoad,
                                                 Real64 &FullSensibleOutput,
-                                                bool HXUnitOn, // Flag to control HX for HXAssisted Cooling Coil
+                                                bool &HXUnitOn, // Flag to control HX for HXAssisted Cooling Coil
```
Comment (rraustad) [inline src/EnergyPlus/UnitarySystem.cc:7792]: This function switches on the HX as needed. This info needed to be passed back out and used in subsequent calls.

File: src/EnergyPlus/UnitarySystem.cc [hunk 2]
```diff
@@ -8549,7 +8549,7 @@ namespace UnitarySystems {
                         if (CoolingLoad) Par[6] = 1.0;
                         Par[7] = 1.0;               // FLAG, 0.0 if latent load, 1.0 if sensible load to be met
                         Par[8] = OnOffAirFlowRatio; // Ratio of compressor ON mass flow rate to AVERAGE mass flow rate over time step
-                        Par[9] = 0.0;               // HXUnitOn is always false for HX
+                        Par[9] = HXUnitOn;          // HXUnitOn turns on or off HX in coil system
```
Comment (rraustad) [inline src/EnergyPlus/UnitarySystem.cc:8552]: Not sure why the HXUnitOn flag was always set to 0 here. Unit tests show it needs to be controlled based on the flag.

File: src/EnergyPlus/UnitarySystem.cc [hunk 3]
```diff
@@ -8881,6 +8881,7 @@ namespace UnitarySystems {
             }
 
             if (MoistureLoad < LatOutputOn && this->m_DehumidControlType_Num == DehumCtrlType::CoolReheat) {
+                HXUnitOn = true; // HX is needed to meet moisture load
```
Comment (rraustad) [inline src/EnergyPlus/UnitarySystem.cc:8884]: And for CoolReheat the HX needs to be turned on when sensible only load control can't meet the latent load.

File: tst/EnergyPlus/unit/HVACHXAssistedCoolingCoil.unit.cc [hunk 4]
```diff
@@ -514,11 +512,110 @@ TEST_F(EnergyPlusFixture, HXAssistCCUnitarySystem_VStest1)
     Qsens_sys = DataLoopNode::Node(InletNode).MassFlowRate *
                 (Psychrometrics::PsyHFnTdbW(DataLoopNode::Node(OutletNode).Temp, MinHumRatio) - Psychrometrics::PsyHFnTdbW(ZoneTemp, MinHumRatio));
 
-    // TODO: FIXME: Need to fix this in future, it is failing now, probably due to object ordering. Unit test failure message below
-    // The difference between DataZoneEnergyDemands::ZoneSysEnergyDemand( ControlZoneNum ).RemainingOutputRequired and Qsens_sys is 1000, which
-    // exceeds 1.0, where DataZoneEnergyDemands::ZoneSysEnergyDemand( ControlZoneNum ).RemainingOutputRequired evaluates to -1000, Qsens_sys evaluates
-    // to 0, and 1.0 evaluates to 1. test model performance
```
Comment (rraustad) [inline tst/EnergyPlus/unit/HVACHXAssistedCoolingCoil.unit.cc:520]: Added multiple tests here to cover all (most?) cases.

**Comment by rraustad** [comment]
Same answer as before these new changes.

image

**Comment by mitchute** [comment]
No need to dwell on this. Those hardcoded flags clearly should have been passed through to subsequent functions rather than being hardcoded. CI is happy and the unit tests look good. I ran ```5Zone_Unitary_HXAssistedCoil``` locally and confirmed that the outlet temps now look reasonable. This issue also cleans up a bunch of out-of-bounds temperature warnings in the ERR file. Good stuff all around. Merging.

**Pull request merged**